docs/guides/contexts_expressions.md

```diff
@@ -0,0 +1,390 @@
+
+As mentioned, a context is ultimately just a mapping between variable names and
+their values. (When evaluating a napari expression with `Expr.eval`, you can
+indeed just pass a `dict` as that mapping.)
+
+Important objects in napari, such as the `Viewer` and the `LayerList` will be
+associated with a `Context` object that tracks the value of various context
+keys.  It is the job of these various objects (i.e. the `Viewer` and the
+`LayerList`) to update the values in their `Contexts` when they change.
+Continuing with the example above, if the user clicks on a 4-dimensional layer,
+the `LayerList` would set the context key `active_layer_ndim` to `4`.  (napari
+would then be able to enable/disable various commands & menus that required a
+specific number of dimensions in the active layer)
+
+### The `Context` class
+
+The napari `Context` class is a subclass of
+[`collections.ChainMap`](https://docs.python.org/3/library/collections.html#collections.ChainMap)
+that also emits events when a key has been modified. `ChainMap` is useful here
+as it allows us to have "sub-contexts" that are children of some parent context.
+Child contexts can access all of the keys of the parent (but not vice-versa).
+For example because a `Viewer` has a `LayerList`, all of the keys in the
+`Viewer` context are available to the `LayerList` context. 
+
+```python
+In [1]: from napari.utils.context import get_context
+
+In [2]: ctx = get_context(viewer.layers)
+
+In [3]: ctx
+Out[3]: Context(
+    {
+    'layers_selection_count': 0,
+    'all_layers_linked': False,
+    'unselected_linked_layers': 0,
+    'active_layer_is_rgb': False,
+    'active_layer_type': None,
+    'only_images_selected': False,
+    'only_labels_selected': False,
+    'active_layer_ndim': None,
+    'active_layer_shape': None,
+    'active_layer_dtype': None,
+    'all_layers_same_shape': False
+    }, 
+    {},  # the viewer context ... (no keys yet)
+    SettingsAwareContext({})
+)
+```
+The "root" context is a special `SettingsAwareContext` that can access keys in
+the global `settings`.  Because contexts are `ChainMaps`, they can all access
+the settings:
+
+```python
+In [4]: ctx['settings.appearance.theme']
+Out[4]: 'dark'
+```
+
+When we evaluate an expression, we can provide it one of these context objects:
+
+```python
+In [5]: expr = LayerListContextKeys.layers_selection_count > 0
+
+In [6]: expr.eval(ctx)
+Out[6]: False
+```
+
+### Updating Contexts
+
+You may be wondering exactly how objects such as `Viewer` and `LayerList` update
+the keys in their contexts. The aforementioned
+[`ContextNamespace`](#contextkey-objects) comes into play here again.  A
+`ContextNamespace` can be instantiated, and bound to a specific `Context`
+instance.
+
+```python
+In [6]: ctx = get_context(viewer.layers)
+
+In [7]: llck = LayerListContextKeys(ctx)
+```
+
+Attributes of an instantiated `ContextNamespace` now act as getters (and
+setters!) of their respective `ContextKey` in the associated `Context`.
```

Comment:
What does acting as a setter mean here? The class attributes of `ContextNamespace` are functions that can be evaluated to generate a value. Does setting mean that the function is re-evaluated and anything dependent on that value is notified of a change?

Reply:
> The class attributes of ContextNamespace are functions that can be evaluated to generate a value.

not quite, the class attributes are _Expr_ objects (specifically `Name` objects, similar to `ast.Name`).  as mentioned in https://github.com/napari/napari/pull/3571#discussion_r744914320 ... these are not primarily functions that can be evaluated.  The fact that they come with a function to generate a value is really more of a "feature" of the `ContextKey` class (but their core functionality is to serve as a expression object that points to a name in some context)

> What does acting as a setter mean here? 

```py
In [25]: from napari.utils.context import LayerListContextKeys

In [26]: ctx = {}

In [27]: llck = LayerListContextKeys(ctx)

In [28]: llck.unselected_linked_layers
Out[28]: 0

In [29]: ctx['unselected_linked_layers']  # the above is shorthand for this
Out[29]: 0

In [30]: llck.unselected_linked_layers = 2  # <--- it's a setter for the value in the bound context

In [31]: ctx['unselected_linked_layers']
Out[31]: 2
```